### PR TITLE
Create talker monster, add ability for spells to use dialog effects/conditions

### PR DESCRIPTION
**Pull request by Ramza13**
#### Summary
Infrastructure "Create talker monster, add ability for spells to use dialog effects/conditions"

#### Purpose of change
Integrate spells and dialog effects/conditions
Create talker monster, setting stage for having conversations with monsters. 

#### Describe the solution
First commit is adding talker monster class.  
Second commit is adding spell type to use dialog effects/conditions.  Change message dialog effect to allow using it on either conversation participant(if a player casts a spell they will be in the npc side of things so will only see messages targeted to the npc).
Fourth commit is adding ability for dialog effects to cast a spell.

#### Describe alternatives you've considered
This creates some overlap in functionality(adding or removing effects for example can now be done two ways in spells).  Decided to tackle that in the future.

#### Testing
Played around with this json
```
{
    "id": "eoc",
    "type": "SPELL",
    "name": { "str": "eoc" },
    "description": "eoc",
    "valid_targets": [ "hostile", "ally", "self", "ground" ],
    "flags": [ "SILENT" ],
    "//": "Listed as a recover energy effect with a negative modifier that decreases with each level of the spell, which makes it cause damage instead.",
    "max_level": 10,
    "shape": "blast",
    "min_range": 1,
    "max_range": 25,
    "effect": "effect_on_condition",
    "effect_str": "EOC_TEST"
  },
  {
    "type": "effect_on_condition",
    "id": "EOC_TEST",
    "effect": [
      {
        "set_queue_effect_on_condition": [
          {
            "id": "EOC_1",
            "condition": { "u_has_var": "npc_foodperson_impressed", "type": "dialogue", "context": "foodperson", "value": "yes" },
            "effect": [ { "npc_message": "Good1" },{ "u_message": "Bad1" } ]
          },
          {
            "id": "EOC_2",
            "condition": { "not": { "u_has_var": "npc_foodperson_impressed", "type": "dialogue", "context": "foodperson", "value": "yes" } },
            "effect": [
              { "u_message": "Bad2" },
              { "npc_message": "Good2" },
              { "u_add_var": "npc_foodperson_impressed", "type": "dialogue", "context": "foodperson", "value": "yes" }
            ]
          }
        ]
      }
    ]
  }
```

For dialog effects casting spells used this to test
```
{
    "type": "effect_on_condition",
    "id": "test_spell",
    "recurrence_min": 1,
    "recurrence_max": 1,
    "effect": [
      { "u_message": "A test happens" },
      { "u_cast_spell": { "id": "mi-go_slaver_beam", "min_level": 3 } }
    ]
  },
```

### CODE REVIEW

**Comment by actual-nh** [comment]
Perhaps dialog effects/conditions controlling special attack use next? Anger and morale as available variables would also be very helpful. Thanks for all your work on this!

**Comment by Ramza13** [comment]
> 
> 
> Perhaps dialog effects/conditions controlling special attack use next? Anger and morale as available variables would also be very helpful. Thanks for all your work on this!

So aren't spells doable as special attacks?  If not then sure, I just assumed this handled that case as well.

**Comment by actual-nh** [comment]
> > Perhaps dialog effects/conditions controlling special attack use next? Anger and morale as available variables would also be very helpful. Thanks for all your work on this!
> 
> So aren't spells doable as special attacks? If not then sure, I just assumed this handled that case as well.

I was referring to other special attacks (e.g., guns). Thanks!

**Comment by Ramza13** [comment]
Just added ability for dialog effects to cast a spell.

**Comment by actual-nh** [comment]
Part of what I'm thinking about is somewhat more sensible behavior for monsters with limited ammunition, something that takes a long time to recharge, or whatever - trigger using a gun with limited bullets (by a more-intelligent feral, as the docs state should exist, for instance) when, say, 2 of low HP, aggression above starting, morale _below_ starting.

**Comment by KorGgenT** [comment]
i don't really like that you're adding a member to spell_type that's only used for a single spell effect. i don't really have the brain space to review this fully, but i expect there to be other options to explore to avoid doing that. perhaps you can leverage some kind of dialogue id in `effect_str` or somesuch?

**Comment by Ramza13** [comment]
> 
> 
> i don't really like that you're adding a member to spell_type that's only used for a single spell effect. i don't really have the brain space to review this fully, but i expect there to be other options to explore to avoid doing that. perhaps you can leverage some kind of dialogue id in `effect_str` or somesuch?

Would changing effect_str to a vector be a better solution?  I'd like to be able to pass multiple string id's though I suppose its not strictly speaking necessary as you can have one eoc call multiple but I figured the more nesting I can avoid the better.

**Comment by Ramza13** [comment]
Just removed the new effects_on_condition parameter and reworked effect_str to optionally be an array and used it instead.

**Comment by KorGgenT** [comment]
> > i don't really like that you're adding a member to spell_type that's only used for a single spell effect. i don't really have the brain space to review this fully, but i expect there to be other options to explore to avoid doing that. perhaps you can leverage some kind of dialogue id in `effect_str` or somesuch?
> 
> Would changing effect_str to a vector be a better solution? I'd like to be able to pass multiple string id's though I suppose its not strictly speaking necessary as you can have one eoc call multiple but I figured the more nesting I can avoid the better.

no, i'd much rather rework all of spells to use inheritance, i'd like to keep effect_str to be a string. i really wish i had more time to examine what's going on here buti haven't had time in the past couple days.

File: src/magic.cpp [hunk 1]
```diff
@@ -348,6 +346,21 @@ void spell_type::load( const JsonObject &jo, const std::string & )
...
+            for( JsonValue jv : jo.get_array( "effect_str" ) ) {
+                effect_str.emplace_back( effect_on_conditions::load_inline_eoc( jv, "" ).c_str() );
+            }
+        } else {
+            mandatory( jo, was_loaded, "effect_str", effect_str );
+        }
+    }
+    if( effect_str.empty() ) {
+        effect_str.push_back( effect_str_default );
+    }
```
Comment (KorGgenT) [inline src/magic.cpp:363]: i don't want the array in here. it's one-off code, it makes the loading code more messy, and you already mentioned that you can do effect_on_condition with a single string by utilizing nested conditions.

**Comment by Ramza13** [comment]
Removed the array bits and fixed an issue with queueing eocs at the same time.

**Comment by actual-nh** [comment]
I am guessing that #47090 could be done using these capabilities?

File: src/npctalk.cpp [hunk 2]
```diff
@@ -2299,19 +2314,31 @@ void talk_effect_fun_t::set_queue_effect_on_condition( const JsonObject &jo,
...
+            if( Creature *creature = d.alpha.get()->get_creature() ) {
+                newDialog.alpha = get_talker_for( creature );
+            } else if( item_location *item = d.alpha.get()->get_item() ) {
+                newDialog.alpha = get_talker_for( item );
+            } else {
+                debugmsg( "Invalid alpha talker." );
+            }
+            if( Creature *creature = d.beta.get()->get_creature() ) {
+                newDialog.beta = get_talker_for( creature );
+            } else if( item_location *item = d.beta.get()->get_item() ) {
```
Comment (kevingranade) [inline src/npctalk.cpp:2334]:
```suggestion
            if( Creature *creature = d.alpha->get_creature() ) {
                newDialog.alpha = get_talker_for( creature );
            } else if( item_location *item = d.alpha->get_item() ) {
                newDialog.alpha = get_talker_for( item );
            } else {
                debugmsg( "Invalid alpha talker." );
            }
            if( Creature *creature = d.beta->get_creature() ) {
                newDialog.beta = get_talker_for( creature );
            } else if( item_location *item = d.beta->get_item() ) {
```

File: tests/json_test.cpp [hunk 3]
```diff
@@ -140,7 +140,7 @@ TEST_CASE( "spell_type handles all members", "[json]" )
             R"("effect":"attack",)"
             R"("shape":"blast",)"
             R"("valid_targets":["none"],)"
-            R"("effect_str":"string",)"
+            R"("effect_str":"[string]",)"
```
Comment (kevingranade) [inline tests/json_test.cpp:143]:
```suggestion
            R"("effect_str":"string",)"
```

**Pull request merged**